File: apis/containers.py
from flask import Blueprint, request, jsonify
import docker

containers_blueprint = Blueprint('container', __name__)
client = docker.DockerClient(base_url='unix://var/run/docker.sock')

# Endpoint to list all Docker containers with detailed information
@containers_blueprint.route('/containers', methods=['GET'])
def list_containers():
    containers = client.containers.list(all=True)
    container_list = []
    for container in containers:
        container_info = {
            'id': container.id,
            'name': container.name,
            'status': container.status,
            'image': container.image.tags,
            'labels': container.labels,
            'network_settings': container.attrs['NetworkSettings'],
            'mounts': container.attrs['Mounts'],
            'created': container.attrs['Created'],
            'state': container.attrs['State'],
            'command': container.attrs['Config']['Cmd']
        }
        container_list.append(container_info)
    return jsonify(container_list)

# Endpoint to create a Docker container
@containers_blueprint.route('/containers/create', methods=['POST'])
def create_container():
    data = request.json
    name = data.get('name')
    image = data.get('image')
    volumes = data.get('volumes')
    user = data.get('user')
    network = data.get('network')
    env = data.get('env')
    ports = data.get('ports')
    command = data.get('command')
    
    container = client.containers.create(
        image,
        name=name,
        volumes=volumes,
        user=user,
        network=network,
        environment=env,
        ports=ports,
        command=command
    )
    return jsonify({'id': container.id, 'name': container.name})

# Endpoint to remove a Docker container
@containers_blueprint.route('/containers/remove/<string:container_id>', methods=['DELETE'])
def remove_container(container_id):
    container = client.containers.get(container_id)
    container.remove(force=True)
    return '', 204

# Endpoint to start a Docker container
@containers_blueprint.route('/containers/start/<string:container_id>', methods=['POST'])
def start_container(container_id):
    container = client.containers.get(container_id)
    container.start()
    return '', 204

# Endpoint to stop a Docker container
@containers_blueprint.route('/containers/stop/<string:container_id>', methods=['POST'])
def stop_container(container_id):
    container = client.containers.get(container_id)
    container.stop()
    return '', 204

# Endpoint to pause a Docker container
@containers_blueprint.route('/containers/pause/<string:container_id>', methods=['POST'])
def pause_container(container_id):
    container = client.containers.get(container_id)
    container.pause()
    return '', 204

# Endpoint to rename a Docker container
@containers_blueprint.route('/containers/rename/<string:container_id>', methods=['POST'])
def rename_container(container_id):
    data = request.json
    new_name = data.get('new_name')
    container = client.containers.get(container_id)
    container.rename(new_name)
    return '', 204

# Endpoint to change user of a Docker container
@containers_blueprint.route('/containers/change-user/<string:container_id>', methods=['POST'])
def change_user(container_id):
    data = request.json
    new_user = data.get('new_user')
    container = client.containers.get(container_id)
    container.update(user=new_user)
    return '', 204

# Endpoint to add a port to a Docker container
@containers_blueprint.route('/containers/add-port/<string:container_id>', methods=['POST'])
def add_port(container_id):
    data = request.json
    port = data.get('port')
    container = client.containers.get(container_id)
    container.update(ports={port: port})
    return '', 204

# Endpoint to remove a port from a Docker container
@containers_blueprint.route('/containers/remove-port/<string:container_id>', methods=['POST'])
def remove_port(container_id):
    data = request.json
    port = data.get('port')
    container = client.containers.get(container_id)
    container.update(ports={port: None})
    return '', 204

# Endpoint to update a port in a Docker container
@containers_blueprint.route('/containers/update-port/<string:container_id>', methods=['POST'])
def update_port(container_id):
    data = request.json
    old_port = data.get('old_port')
    new_port = data.get('new_port')
    container = client.containers.get(container_id)
    container.update(ports={old_port: None, new_port: new_port})
    return '', 204

# Endpoint to add a volume to a Docker container
@containers_blueprint.route('/containers/add-volume/<string:container_id>', methods=['POST'])
def add_volume(container_id):
    data = request.json
    volume = data.get('volume')
    container = client.containers.get(container_id)
    container.update(volumes={volume: volume})
    return '', 204

# Endpoint to remove a volume from a Docker container
@containers_blueprint.route('/containers/remove-volume/<string:container_id>', methods=['POST'])
def remove_volume(container_id):
    data = request.json
    volume = data.get('volume')
    container = client.containers.get(container_id)
    container.update(volumes={volume: None})
    return '', 204

# Endpoint to get logs of a Docker container
@containers_blueprint.route('/containers/logs/<container_name>', methods=['GET'])
def get_logs(container_name):
    """Retrieve logs from a container and return them as a list."""
    client = docker.from_env()
    try:
        # Get container
        container = client.containers.get(container_name)
        
        # Retrieve logs
        logs = container.logs(tail=1000).decode('utf-8')  # Tail 1000 lines or adjust as needed

        # Split logs into lines and return as a list
        log_lines = logs.splitlines()

        return jsonify(log_lines), 200

    except docker.errors.NotFound:
        return jsonify({'error': 'Container not found'}), 404
    except Exception as e:
        return jsonify({'error': str(e)}), 500

# Endpoint to update restart policy of a Docker container
@containers_blueprint.route('/containers/update-restart/<string:container_id>', methods=['POST'])
def update_restart(container_id):
    data = request.json
    restart_policy = data.get('restart_policy')
    container = client.containers.get(container_id)
    container.update(restart_policy=restart_policy)
    return '', 204

# Endpoint to add a command to a Docker container
@containers_blueprint.route('/containers/add-command/<string:container_id>', methods=['POST'])
def add_command(container_id):
    data = request.json
    command = data.get('command')
    container = client.containers.get(container_id)
    container.update(command=command)
    return '', 204

# Endpoint to remove a command from a Docker container
@containers_blueprint.route('/containers/remove-command/<string:container_id>', methods=['POST'])
def remove_command(container_id):
    container = client.containers.get(container_id)
    container.update(command=None)
    return '', 204

# Endpoint to update a command in a Docker container
@containers_blueprint.route('/containers/update-command/<string:container_id>', methods=['POST'])
def update_command(container_id):
    data = request.json
    new_command = data.get('new_command')
    container = client.containers.get(container_id)
    container.update(command=new_command)
    return '', 204

# Endpoint to add an environment variable to a Docker container
@containers_blueprint.route('/containers/add-environment/<string:container_id>', methods=['POST'])
def add_environment(container_id):
    data = request.json
    env = data.get('env')
    container = client.containers.get(container_id)
    container.update(environment={**container.attrs['Config']['Env'], **env})
    return '', 204

# Endpoint to remove an environment variable from a Docker container
@containers_blueprint.route('/containers/remove-environment/<string:container_id>', methods=['POST'])
def remove_environment(container_id):
    data = request.json
    env = data.get('env')
    container = client.containers.get(container_id)
    env_vars = container.attrs['Config']['Env']
    env_vars = [var for var in env_vars if var not in env]
    container.update(environment=env_vars)
    return '', 204

# Endpoint to update an environment variable in a Docker container
@containers_blueprint.route('/containers/update-environment/<string:container_id>', methods=['POST'])
def update_environment(container_id):
    data = request.json
    old_env = data.get('old_env')
    new_env = data.get('new_env')
    container = client.containers.get(container_id)
    env_vars = container.attrs['Config']['Env']
    env_vars = [new_env if var == old_env else var for var in env_vars]
    container.update(environment=env_vars)
    return '', 204

File: frontend/dev/api/containers.ts
const API_BASE_URL = 'http://192.168.1.50:5000/api'; // API temel URL'si

import { Container } from "@/models";
import toast from 'react-hot-toast';


// Tüm Docker konteynerlarını listele
export const listContainers = async (): Promise<Container[]> => {
  const response = await fetch(`${API_BASE_URL}/containers`);
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Docker konteyner oluştur
export const createContainer = async (data: any): Promise<any> => {
  const response = await fetch(`${API_BASE_URL}/containers/create`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(data),
  });
  if (!response.ok) {
    toast.error('Container could not be created');
    throw new Error('Network response was not ok');
  }
  toast.success('Container created successfully');
  return response.json();
};

// Docker konteyner sil
export const removeContainer = async (containerId: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/remove/${containerId}`, {
    method: 'DELETE',
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner başlat
export const startContainer = async (containerId: string): Promise<void> => {
  const startContainerRequest = async (containerId: string): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/containers/start/${containerId}`, {
      method: 'POST',
    });
    if (!response.ok) {
      throw new Error('Network response was not ok');
    }
  }
  await toast.promise(startContainerRequest(containerId), {
    loading: 'Starting container...',
    success: 'Container started successfully',
    error: 'Container could not be started',
  });
};

// Docker konteyner durdur
export const stopContainer = async (containerId: string): Promise<void> => {
  const stopContainerRequest = async (containerId: string): Promise<void> => {
    const response = await fetch(`${API_BASE_URL}/containers/stop/${containerId}`, {
      method: 'POST',
    });
    if (!response.ok) {
   
      throw new Error('Network response was not ok');
    }
  }

  await toast.promise(stopContainerRequest(containerId), {  
    loading: 'Stopping container...',
    success: 'Container stopped successfully',
    error: 'Container could not be stopped',
  });
  
};

// Docker konteyner duraklat
export const pauseContainer = async (containerId: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/pause/${containerId}`, {
    method: 'POST',
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner adını değiştir
export const renameContainer = async (containerId: string, newName: string): Promise<void> => {
  const renameContainerRequest = async (containerId: string, newName: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/rename/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ new_name: newName }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  }
  await toast.promise(renameContainerRequest(containerId, newName), {
    loading: 'Renaming container...',
    success: 'Container renamed successfully',
    error: 'Container could not be renamed',
  });
};

// Docker konteyner kullanıcısını değiştir
export const changeUser = async (containerId: string, newUser: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/change-user/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ new_user: newUser }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner portu ekle
export const addPort = async (containerId: string, port: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/add-port/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ port }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner portunu kaldır
export const removePort = async (containerId: string, port: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/remove-port/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ port }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner portunu güncelle
export const updatePort = async (containerId: string, oldPort: string, newPort: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/update-port/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ old_port: oldPort, new_port: newPort }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner hacmi ekle
export const addContainerVolume = async (containerId: string, volume: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/add-volume/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ volume }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner hacmini kaldır
export const removeContainerVolume = async (containerId: string, volume: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/remove-volume/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ volume }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner loglarını al
export const getLogs = async (containerName: string): Promise<string[]> => {
  const response = await fetch(`${API_BASE_URL}/containers/logs/${containerName}`);
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
  return response.json();
};

// Docker konteyner yeniden başlatma politikasını güncelle
export const updateRestartPolicy = async (containerId: string, restartPolicy: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/update-restart/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ restart_policy: restartPolicy }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner komutunu ekle
export const addCommand = async (containerId: string, command: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/add-command/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ command }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner komutunu kaldır
export const removeCommand = async (containerId: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/remove-command/${containerId}`, {
    method: 'POST',
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner komutunu güncelle
export const updateCommand = async (containerId: string, newCommand: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/update-command/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ new_command: newCommand }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner ortam değişkeni ekle
export const addEnvironment = async (containerId: string, env: Record<string, string>): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/add-environment/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ env }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner ortam değişkenini kaldır
export const removeEnvironment = async (containerId: string, env: string[]): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/remove-environment/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ env }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};

// Docker konteyner ortam değişkenini güncelle
export const updateEnvironment = async (containerId: string, oldEnv: string, newEnv: string): Promise<void> => {
  const response = await fetch(`${API_BASE_URL}/containers/update-environment/${containerId}`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ old_env: oldEnv, new_env: newEnv }),
  });
  if (!response.ok) {
    throw new Error('Network response was not ok');
  }
};
